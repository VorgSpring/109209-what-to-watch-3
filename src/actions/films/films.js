import {
  LOAD_FILMS_REQUEST,
  LOAD_FILMS_SUCCESS,
  LOAD_FILMS_ERROR
} from '../../constants/actions-type';

export const loadFilmsRequest = () => ({
  type: LOAD_FILMS_REQUEST
});

export const loadFilmsSuccess = (films) => ({
  type: LOAD_FILMS_SUCCESS,
  payload: films
});

export const loadFilmsError = (error) => ({
  type: LOAD_FILMS_ERROR,
  payload: error
});

