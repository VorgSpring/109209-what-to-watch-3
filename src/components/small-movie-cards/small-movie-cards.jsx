import React from 'react';
import PropTypes from 'prop-types';
import {SmallMovieCard} from '../small-movie-card/small-movie-card.jsx';

export const SmallMovieCards = ({films}) => (
  <div className="catalog__movies-list">
    {films.map((film) => (
      <SmallMovieCard
        key={film.id}
        film={film} />
    ))}
  </div>
);

SmallMovieCards.propTypes = {
  films: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        previewImage: PropTypes.string.isRequired,
        preview: PropTypes.oneOfType([
          PropTypes.array,
          PropTypes.string
        ]).isRequired,
      })
  ).isRequired
};
