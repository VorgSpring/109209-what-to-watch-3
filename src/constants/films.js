export const MAX_COUNT_SIMILAR_FILMS = 4;

export const ALL_GENRES_TYPE = `All genres`;
