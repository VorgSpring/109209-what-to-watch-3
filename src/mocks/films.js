export const films = [
  {
    id: 1,
    bgImage: `/bg-image`,
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Crime`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah-blah`
    ],
    runTime: 123
  },
  {
    id: 2,
    bgImage: `/bg-image`,
    name: `Bohemian Rhapsody`,
    poster: `img/bohemian-rhapsody.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Romance`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 3,
    bgImage: `/bg-image`,
    name: `Macbeth`,
    poster: `img/macbeth.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Horror`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 4,
    bgImage: `/bg-image`,
    name: `Aviator`,
    poster: `img/aviator.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Dramas`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 5,
    bgImage: `/bg-image`,
    bgColor: `#fff`,
    name: `We need to talk about Kevin`,
    poster: `img/we-need-to-talk-about-kevin.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Documentary`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ],
    runTime: 123
  },
];

/* eslint-disable camelcase */
export const filmByServer = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  poster_image: `img/the-grand-budapest-hotel-poster.jpg`,
  preview_image: `img/the-grand-budapest-hotel.jpg`,
  background_image: `img/the-grand-budapest-hotel-bg.jpg`,
  background_color: `#ffffff`,
  video_link: `https://some-link`,
  preview_video_link: `https://some-link`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  rating: 8.9,
  scores_count: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  run_time: 99,
  genre: `Comedy`,
  released: 2014,
  is_favorite: false,
};
