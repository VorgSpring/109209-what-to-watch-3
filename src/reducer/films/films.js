import {
  LOAD_FILMS_REQUEST,
  LOAD_FILMS_SUCCESS,
  LOAD_FILMS_ERROR,
  CHANGE_FAVORITES_SUCCESS
} from '../../constants/actions-type';

const initState = {
  isLoading: false,
  isError: false,
  films: null
};

export default (state = initState, action) => {
  switch (action.type) {
    case (LOAD_FILMS_REQUEST):
      return Object.assign(
          {}, state, {
            isLoading: true,
            isError: false
          }
      );

    case (LOAD_FILMS_SUCCESS):
      return Object.assign(
          {}, state, {
            isLoading: false,
            films: action.payload
          }
      );

    case (CHANGE_FAVORITES_SUCCESS):
      return Object.assign(
          {}, state, {
            films: state.films.map((film) => {
              if (film.id === action.payload.id) {
                return action.payload;
              }
              return film;
            })
          }
      );

    case (LOAD_FILMS_ERROR):
      return Object.assign(
          {}, state, {
            isLoading: false,
            isError: true
          }
      );
    default:
      return state;
  }
};
