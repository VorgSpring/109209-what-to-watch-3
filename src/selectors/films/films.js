import {createSelector} from 'reselect';
import {
  getFilterGenreSelector
} from '../filters/filters';
import {
  MAX_COUNT_SIMILAR_FILMS,
  ALL_GENRES_TYPE
} from '../../constants/films';

export const getFilmsLoadedStatusSelector = (state) =>
  state.films && state.films.isLoaded || false;

export const getFilmsSelector = (state) =>
  state.films && state.films.films || [];

export const getFilmIdSelector = (_, props) => {
  const id = parseInt(
      props && props.match && props.match.params && props.match.params.id, 10
  );

  return isNaN(id) ? null : id;
};

export const getFiltratedFilmsSelector = createSelector(
    getFilmsSelector,
    getFilterGenreSelector,
    (films, filter) => {
      if (filter === ALL_GENRES_TYPE) {
        return films;
      }

      return films.filter(({genre}) => (
        genre === filter
      ));
    }
);

export const getFilmByIdSelector = createSelector(
    getFilmsSelector,
    getFilmIdSelector,
    (films, id) => {
      if (!films || !id) {
        return null;
      }

      return films.find((film) => film.id === id);
    }
);

export const getSimilarFilmsSelector = createSelector(
    getFilmsSelector,
    getFilmByIdSelector,
    (films, film) => {
      if (!film) {
        return [];
      }

      return films.filter(({genre, id}) => (
        genre === film.genre && id !== film.id
      )).slice(0, MAX_COUNT_SIMILAR_FILMS);
    }
);
