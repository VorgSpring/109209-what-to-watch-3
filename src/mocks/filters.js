import {ALL_GENRES_TYPE} from '../constants/films';

export const filters = {
  genre: ALL_GENRES_TYPE.ALL
};
