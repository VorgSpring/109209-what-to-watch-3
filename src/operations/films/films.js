import {
  loadFilmsRequest,
  loadFilmsSuccess,
  loadFilmsError
} from '../../actions/films/films';
import {ApiPaths} from '../../constants/api';
import {preparedFilmsData} from '../../helpers/prepared-films-data/prepared-films-data';

export const loadFilms = () => (dispatch, _getState, api) => {
  dispatch(loadFilmsRequest());

  return api.get(ApiPaths.FILMS)
    .then((response) => {
      const films = preparedFilmsData(response.data);
      dispatch(loadFilmsSuccess(films));
    })
    .catch((e) => {
      const {error} = e.response.data;
      dispatch(loadFilmsError({error}));
    });
};
