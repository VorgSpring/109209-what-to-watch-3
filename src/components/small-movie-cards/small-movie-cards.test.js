import React from 'react';
import {SmallMovieCards} from './small-movie-cards.jsx';
import renderer from 'react-test-renderer';
import {films} from '../../mocks/films';

jest.mock(`../small-movie-card/small-movie-card.jsx`, () => ({
  SmallMovieCard() {
    return <small-movie-card />;
  }
}));

describe(`SmallMovieCards`, () => {
  it(`should renders correctly`, () => {
    const tree = renderer.create(
        <SmallMovieCards
          films={films}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
