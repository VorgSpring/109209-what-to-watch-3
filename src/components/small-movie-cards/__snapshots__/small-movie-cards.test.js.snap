// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallMovieCards should renders correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <small-movie-card />
  <small-movie-card />
  <small-movie-card />
  <small-movie-card />
  <small-movie-card />
  <small-movie-card />
</div>
`;
